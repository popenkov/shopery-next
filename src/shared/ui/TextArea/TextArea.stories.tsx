import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import { TextArea } from './TextArea';

const meta = {
  title: 'shared/TextArea',
  component: TextArea,
  parameters: {
    layout: 'centered',
  },
  args: {
    onBlur: fn(),
    onChange: fn(),
    onFocus: fn(),
  },
  tags: ['autodocs'],
} satisfies Meta<typeof TextArea>;

export default meta;
type TStory = StoryObj<typeof meta>;

export const TextAreaComponent: TStory = {
  decorators: [
    (Story) => (
      <div style={{ maxWidth: '400px' }}>
        <Story />
      </div>
    ),
  ],
  args: { placeholder: 'name' },
};

export const TextAreaFilledComponent: TStory = {
  decorators: [
    (Story) => (
      <div style={{ maxWidth: '400px' }}>
        <Story />
      </div>
    ),
  ],
  args: {
    placeholder: 'name',
  },
};
