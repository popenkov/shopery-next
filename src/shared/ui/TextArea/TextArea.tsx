'use client';

import { ChangeEvent, InputHTMLAttributes, forwardRef, memo, useId, useState } from 'react';
import cn from 'classnames';

import cls from './TextArea.module.scss';
import { Text } from '../Text';

type HTMLTextAreaProps = Omit<
  InputHTMLAttributes<HTMLTextAreaElement>,
  'value' | 'onChange' | 'readOnly'
>;

interface Props extends HTMLTextAreaProps {
  className?: string;
  label?: string;
  errorText?: string;
  hasValidation?: boolean;
  value?: string | number;
  onChange?: (evt: ChangeEvent<HTMLTextAreaElement>) => void;
}

export const TextArea = memo(
  forwardRef<HTMLTextAreaElement, Props>((props, ref) => {
    const {
      className,
      value,
      label,
      errorText,
      hasValidation = true,
      onChange,
      ...otherProps
    } = props;

    const [isInvalid, setIsInvalid] = useState(true);

    const uniqueId = useId();

    return (
      <>
        <div className={cn(cls.inputWrapper, className)}>
          {label && <label htmlFor={uniqueId}>{label}</label>}
          <textarea
            id={uniqueId}
            ref={ref}
            value={value}
            onChange={(evt) => onChange?.(evt)}
            className={cls.input}
            {...otherProps}
          />
        </div>
        {/* todo */}
        {/* {isInvalid && hasValidation && <span>error</span>} */}
        {errorText && (
          <Text variant="body_s" weight="medium" className={cls.errorMessage} as="span">
            {errorText}
          </Text>
        )}
      </>
    );
  }),
);

TextArea.displayName = 'TextArea';
