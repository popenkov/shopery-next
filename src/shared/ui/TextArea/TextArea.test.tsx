import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { TextArea } from './TextArea';

describe('TextArea component', () => {
  it('renders correctly', () => {
    const { container } = render(<TextArea />);
    expect(container).toMatchSnapshot();
  });

  it('renders correct class names', () => {
    const { container } = render(<TextArea className="custom-class" />);
    expect(container.querySelector('.inputWrapper')).toBeInTheDocument();
    expect(container.querySelector('.input')).toBeInTheDocument();
  });

  it('renders label', () => {
    const { getByText } = render(<TextArea label="Label" />);
    expect(getByText('Label')).toBeInTheDocument();
  });

  it('renders error message', () => {
    const { getByText } = render(<TextArea errorText="Error message" />);
    expect(getByText('Error message')).toBeInTheDocument();
  });

  it('renders textarea with correct id', () => {
    const { getByLabelText } = render(<TextArea label="Label" />);
    const textarea = getByLabelText('Label');
    expect(textarea).toHaveAttribute('id');
  });

  it('renders textarea with correct class names', () => {
    const { container } = render(<TextArea className="custom-class" />);
    const textareaContainer = container.querySelector('.inputWrapper');
    expect(textareaContainer).toHaveClass('custom-class');
  });
});
