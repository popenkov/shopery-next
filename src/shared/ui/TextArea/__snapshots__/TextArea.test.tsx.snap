// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea component renders correctly 1`] = `
<div>
  <div
    class="inputWrapper"
  >
    <textarea
      class="input"
      id=":r0:"
    />
  </div>
</div>
`;
